File: backend/app/services/portfolio.py
"""Portfolio valuation, P&L math, and snapshot recording."""

from __future__ import annotations

import asyncio
import sqlite3

from app.db import DEFAULT_USER_ID, Position
from app.db.repositories import positions as positions_repo
from app.db.repositories import profile as profile_repo
from app.db.repositories import snapshots as snapshots_repo
from app.market import PriceCache
from app.schemas import Portfolio, PortfolioHistory, PositionValue, SnapshotPoint
from app.state import MarketContext


def _percent(numerator: float, denominator: float) -> float:
    """Percentage of numerator over denominator, 0.0 when the base is zero."""
    if denominator == 0:
        return 0.0
    return round(numerator / denominator * 100, 2)


def price_for(cache: PriceCache, position: Position) -> float | None:
    """Latest cached price for a holding, or None when the ticker is untracked."""
    return cache.get_price(position.ticker)


def value_position(
    position: Position,
    price: float | None,
    positions_value: float,
) -> PositionValue:
    """Value a single holding. An untracked ticker is held at its average cost."""
    effective = price if price is not None else position.avg_cost
    market_value = round(position.quantity * effective, 2)
    cost_basis = round(position.quantity * position.avg_cost, 2)
    pnl = round(market_value - cost_basis, 2)
    return PositionValue(
        ticker=position.ticker,
        quantity=position.quantity,
        avg_cost=round(position.avg_cost, 4),
        cost_basis=cost_basis,
        current_price=price,
        market_value=market_value,
        unrealized_pnl=pnl,
        unrealized_pnl_percent=_percent(pnl, cost_basis),
        weight=_percent(market_value, positions_value),
    )


def build_portfolio(
    cache: PriceCache,
    user_id: str = DEFAULT_USER_ID,
    conn: sqlite3.Connection | None = None,
) -> Portfolio:
    """Compute the full portfolio valuation. Synchronous."""
    cash = profile_repo.get_cash_balance(user_id=user_id, conn=conn)
    holdings = positions_repo.list_positions(user_id=user_id, conn=conn)
    priced = [(p, price_for(cache, p)) for p in holdings]

    positions_value = round(
        sum(p.quantity * (price if price is not None else p.avg_cost) for p, price in priced), 2
    )
    values = [value_position(p, price, positions_value) for p, price in priced]
    cost_basis = round(sum(v.cost_basis for v in values), 2)
    unrealized = round(positions_value - cost_basis, 2)

    return Portfolio(
        cash_balance=round(cash, 2),
        positions=values,
        positions_value=positions_value,
        total_value=round(cash + positions_value, 2),
        unrealized_pnl=unrealized,
        unrealized_pnl_percent=_percent(unrealized, cost_basis),
    )


def record_snapshot(
    cache: PriceCache,
    user_id: str = DEFAULT_USER_ID,
    conn: sqlite3.Connection | None = None,
) -> float:
    """Value the portfolio and append a snapshot row. Returns the total value."""
    portfolio = build_portfolio(cache, user_id=user_id, conn=conn)
    snapshots_repo.record_snapshot(portfolio.total_value, user_id=user_id, conn=conn)
    return portfolio.total_value


def build_history(
    limit: int | None = None,
    user_id: str = DEFAULT_USER_ID,
    conn: sqlite3.Connection | None = None,
) -> PortfolioHistory:
    """Load portfolio value snapshots, oldest first. Synchronous."""
    rows = snapshots_repo.list_snapshots(limit=limit, user_id=user_id, conn=conn)
    return PortfolioHistory(
        snapshots=[
            SnapshotPoint(total_value=row.total_value, recorded_at=row.recorded_at) for row in rows
        ]
    )


async def get_portfolio(market: MarketContext, user_id: str = DEFAULT_USER_ID) -> Portfolio:
    """Async wrapper over build_portfolio for use in request handlers."""
    return await asyncio.to_thread(build_portfolio, market.cache, user_id)


async def get_history(
    limit: int | None = None,
    user_id: str = DEFAULT_USER_ID,
) -> PortfolioHistory:
    """Async wrapper over build_history for use in request handlers."""
    return await asyncio.to_thread(build_history, limit, user_id)

File: frontend/src/lib/portfolio.ts
import type { Portfolio, PriceMap } from "@/lib/types";

/** A position revalued against the live stream rather than the fetch snapshot. */
export interface LivePosition {
  ticker: string;
  quantity: number;
  avgCost: number;
  currentPrice: number;
  marketValue: number;
  costBasis: number;
  unrealizedPnl: number;
  unrealizedPnlPercent: number;
  /** Share of total portfolio value, 0-1. Sizes the heatmap rectangles. */
  weight: number;
}

export interface LivePortfolio {
  cash: number;
  positions: LivePosition[];
  positionsValue: number;
  totalValue: number;
  unrealizedPnl: number;
  unrealizedPnlPercent: number;
}

export const EMPTY_PORTFOLIO: LivePortfolio = {
  cash: 0,
  positions: [],
  positionsValue: 0,
  totalValue: 0,
  unrealizedPnl: 0,
  unrealizedPnlPercent: 0,
};

/**
 * Revalue a fetched portfolio at live prices. The backend snapshot supplies
 * quantity, average cost and cash; the price stream supplies everything else,
 * so the header total ticks with the market between fetches.
 */
export function valuePortfolio(
  portfolio: Portfolio | null,
  prices: PriceMap,
): LivePortfolio {
  if (!portfolio) return EMPTY_PORTFOLIO;

  const positions = portfolio.positions.map((position) => {
    const currentPrice =
      prices[position.ticker]?.price ?? position.current_price ?? position.avg_cost;
    const marketValue = currentPrice * position.quantity;
    const costBasis = position.cost_basis;
    const unrealizedPnl = marketValue - costBasis;
    return {
      ticker: position.ticker,
      quantity: position.quantity,
      avgCost: position.avg_cost,
      currentPrice,
      marketValue,
      costBasis,
      unrealizedPnl,
      unrealizedPnlPercent: costBasis === 0 ? 0 : (unrealizedPnl / costBasis) * 100,
      weight: 0,
    };
  });

  const positionsValue = positions.reduce((sum, p) => sum + p.marketValue, 0);
  const costBasis = positions.reduce((sum, p) => sum + p.costBasis, 0);
  const unrealizedPnl = positionsValue - costBasis;
  const totalValue = portfolio.cash_balance + positionsValue;

  for (const position of positions) {
    position.weight = totalValue === 0 ? 0 : position.marketValue / totalValue;
  }
  positions.sort((a, b) => b.marketValue - a.marketValue);

  return {
    cash: portfolio.cash_balance,
    positions,
    positionsValue,
    totalValue,
    unrealizedPnl,
    unrealizedPnlPercent: costBasis === 0 ? 0 : (unrealizedPnl / costBasis) * 100,
  };
}
